### ISSUE TITLE
twisted.web2.stream readline size argument

### ISSUE THREAD

**Issue by twisted-trac** [twisted/twisted#11006]
|<img alt="mkerrin's avatar" src="https://avatars.githubusercontent.com/u/0?s=50" width="50" height="50">| mkerrin reported|
|-|-|
|Trac ID|trac#1451|
|Type|enhancement|
|Created|2006-01-26 20:05:53Z|

Attachments:

* [twisted.web2.diff](https://raw.githubusercontent.com/twisted/twistedmatrix.com-trac-attachments/trunk/ticket/6f7/6f75179a226fff9856a84f6b4ae3e2c6aaa86a74/e6fdd9279787ea5ea0bd8d668897a0cdd8302f93.diff) (7761 bytes) - added by mkerrin on 2006-01-26 20:05:41Z - 

<details><summary>Searchable metadata</summary>

```
trac-id__1451 1451
type__enhancement enhancement
reporter__mkerrin mkerrin
priority__high high
milestone__ 
branch__ 
branch_author__ 
status__closed closed
resolution__fixed fixed
component__web2 web2
keywords__ 
time__1138305953000000 1138305953000000
changetime__1160882895000000 1160882895000000
version__ 
owner__ 
cc__jknight cc__mkerrin cc__exarkun
```
</details>

**Comment by twisted-trac** [comment]
|<img alt="mkerrin's avatar" src="https://avatars.githubusercontent.com/u/0?s=50" width="50" height="50">|mkerrin commented|
|-|-|

```
#!html
<pre>
The twisted.web2.wsgi.InputStream objects readline method doesn't accept a size
arguement making it none file-like.

I have attached a patch for the files wsgi.py and stream.py in order to support
this. Also included are tests for this feature. My only issue with the patch is
that it is slightly complicated but then again so is twisted.web2.stream

If accepted could I please request that this bug fix be back ported to the 2.1.x
 branch. The generated the patch against the trunk but it should easily apply to
this branch.

Thanks a million,

Michael Kerrin - michael.kerrin@...
</pre>
```

**Comment by twisted-trac** [comment]
|[<img alt="jyknight's avatar" src="https://avatars.githubusercontent.com/u/873903?s=50" width="50" height="50">](https://github.com/jyknight)|@jyknight commented|
|-|-|

```
#!html
<pre>
Thanks for the nicely done patch complete with tests! I'd like to note, however, 
that according to: http://www.python.org/peps/pep-0333.html

"""The optional "size" argument to readline() is not supported, as it may be 
complex for server authors to implement, and is not often used in practice."""

So, while much of this lower level bits of this patch could be useful in any 
case, I'm not sure if you really want me to add an extension to the wsgi 
interface that would make applications using it non-portable?
</pre>
```

**Comment by twisted-trac** [comment]
|<img alt="mkerrin's avatar" src="https://avatars.githubusercontent.com/u/0?s=50" width="50" height="50">|mkerrin commented|
|-|-|

```
#!html
<pre>
Sorry for the delay in getting back.

I never noticed that in the specification before, I just just saw that the input
stream should be a file-like object. The reason I started looking at this is
because of the bug http://www.zope.org/Collectors/Zope3-dev/535 (traceback of
someone trying to call readline with a size argument in Zope3) but I have run
into some nasty problems along the way like inconsistencies between cStringIO
and StringIO readline (see
http://sourceforge.net/tracker/index.php?func=detail&amp;aid=1416477&amp;group_id=5470&amp;atid=105470)
which makes me think that the readline method should just be deprecated, and the
wsgi people are correct. Also it seems that the person who creating the Zope bug
was using a patched version of standard cgi.py module. After thinking about it a
lot I think the cgi module from the standard python library, is not quite up to
scratch. It shouldn't use readline but the read method instead, has you can need
up reading a lot a data into memory with the readline method. But this is for an
other day.

So I agree with you that this should probable not be added to the wsgi code but
parts of the patch that apply to the stream and test_stream modules may provide
some usefull utility functionality to some one somewhere done the line. I will
leave it up to you to decide if stream / test_stream part is worth inclusion.

Thanks.
</pre>
```

**Comment by twisted-trac** [comment]
|[<img alt="jyknight's avatar" src="https://avatars.githubusercontent.com/u/873903?s=50" width="50" height="50">](https://github.com/jyknight)|@jyknight commented|
|-|-|

Damnit, used the wrong text field for my comment.

What I meant to say is:
Cleaning up: reclassifying as low prio enhancement.

**Comment by twisted-trac** [comment]
|<img alt="mgedmin's avatar" src="https://avatars.githubusercontent.com/u/0?s=50" width="50" height="50"><a name="note_6"></a>|mgedmin commented|
|-|-|

Apparently Python 2.4.4 will incorporate the cgi.py patch from http://sourceforge.net/tracker/index.php?func=detail&aid=1112549&group_id=5470&atid=105470 and if you now try to pass a twisted.web2.wsgi.InputStream object to cgi.FieldStorage in a Python 2.4.4 prerelease, you'll get an exception.

Zope 3's HTTPInputStream has a temporary workaround (since today): it accepts a 'size' argument to readline() and then ignores it, because passing it on to twisted.web2.wsgi.InputStream would cause an error.

I am not a language lawyer, but http://docs.python.org/lib/bltin-file-objects.html says that file objects have a readline() method that takes one optional 'size' argument.  The difference in the descriptions of 'readline' and 'readlines' makes me think that while readlines() is allowed to ignore its sizehint arg, readline() shouldn't do that.

**Comment by twisted-trac** [comment]
|[<img alt="jyknight's avatar" src="https://avatars.githubusercontent.com/u/873903?s=50" width="50" height="50">](https://github.com/jyknight)<a name="note_7"></a>|@jyknight set status to `assigned`|
|-|-|

Given that cgi.py now uses the size argument, and that nobody seems to be making a strong decision whether that's wrong or wsgi spec is, I'm going to go ahead and merge this patch, despite it being not part of the wsgi spec. Thanks.

**Comment by twisted-trac** [comment]
|[<img alt="jyknight's avatar" src="https://avatars.githubusercontent.com/u/873903?s=50" width="50" height="50">](https://github.com/jyknight)<a name="note_8"></a>|@jyknight removed owner<br>@jyknight set status to `new`|
|-|-|

This and [#2166](https://github.com/twisted/twisted/issues/2166), [#2170](https://github.com/twisted/twisted/issues/2170) fixed in [branches/readline-1451,](branches/readline-1451,) needs reviewz.

**Comment by twisted-trac** [comment]
|[<img alt="exarkun's avatar" src="https://avatars.githubusercontent.com/u/254565?s=50" width="50" height="50">](https://github.com/exarkun)|@exarkun set owner to @exarkun<br>@exarkun set status to `assigned`|
|-|-|

**Comment by twisted-trac** [comment]
|[<img alt="exarkun's avatar" src="https://avatars.githubusercontent.com/u/254565?s=50" width="50" height="50">](https://github.com/exarkun)<a name="note_10"></a>|@exarkun set owner to @jyknight<br>@exarkun set status to `new`|
|-|-|

Basically looks good.  Just a few minor comments, in order of decreasing importance:

 * what is the expected/intended behavior for readline with a size hint which falls after the \r of a newline, but before the \n?  could you add a test for this case?
 * existence of size parameter to the wsgi.InputStream readline method merits a brief comment at least, since it isn't part of the wsgi spec.  wsgi.InputStream's readline, readlines, and __iter__ could all do with docstrings as well.
 * test_readInputLineSize and test_readInputLineSizeNegZero both use __iadd__ which is technically against the coding standard.
 * Can you reformat the multiline test method docstrings to look like the stream.BufferedStream.readline readline method's docstring?
 * Is anything going to interpolate the message of the ValueError raised by BufferedStream.readline?
 * Consistent whitespace around assignment would be nice :)

The new test coverage is good.  I think the wsgi-specific parts could be a little more unit-y, but there is also something to be said for exercising the whole stack, and it isn't causing any noticably slowdown in this case so I suppose it's fine.

As long as at least the first three bullet points are addressed, feel free just to merge without re-review.

**Comment by twisted-trac** [comment]
|<img alt="foom's avatar" src="https://avatars.githubusercontent.com/u/0?s=50" width="50" height="50"><a name="note_11"></a>|foom set status to `closed`|
|-|-|

(In [18460]) Fix wsgi.input's readline, readlines, and __iter__ methods. In detail:

 * Remove BufferedStream.readline maxLength argument.
 * Add size argument to wsgi.input's readline(). This is not part of the WSGI spec, but code
   such as python stdlib cgi.py require it anyways. ([#11006](https://github.com/twisted/twisted/issues/11006))
 * Add BufferedStream.readline size argument, to enable above.

 * Fix wsgi.readlines() and wsgi.readline() to not add extraneous delimiters. ([[#2170](https://github.com/twisted/twisted/issues/2170)](https://github.com/twisted/twisted/issues/2170))
 * Make BufferedStream.readline return delimiter as part of the result, to enable above.

 * Add an __iter__ method to wsgi.input, which is required by the spec. ([[#2166](https://github.com/twisted/twisted/issues/2166)](https://github.com/twisted/twisted/issues/2166))

 * Change fileupload's use of BufferedStream.readline to conform with above changes.

Also, add some docstrings.

Author: mkerrin, jknight
Reviewer: exarkun
Merges branch: readline-1451
Fixes [#11006](https://github.com/twisted/twisted/issues/11006)
Fixes #2166
Fixes #2170

**Comment by twisted-trac** [comment]
|<img alt="Automation's avatar" src="https://avatars.githubusercontent.com/u/0?s=50" width="50" height="50">|Automation removed owner|
|-|-|
